Introduction to Free Monads/article.md

```diff
@@ -0,0 +1,702 @@
+(f >=> g) >=> h = f >=> (g >=> h) -- associativity
+return >=> f = f -- left identity
+f >=> return = f -- right identity
+```
+
+That is why a monad is indeed a monoid: it satisfies the same laws, and the "values" are of the type `Functor f => a -> f b`.
+Note, though, that this is not the category of endofunctors.
+However, but there exists a natural transformation to the category of endofunctors[^endofunctors].
+
+[^endofunctors]: One could alternatively define monads via `join :: m (m a) -> m a` instead of `>>=`, which would make it a little more evident that monads are monoids over endofunctors, but to explain this properly, we would need to use a bit more of category theory jargon.
+    There is an excellent [r/math post][endofunctors] on the subject.
+
+    [endofunctors]: https://www.reddit.com/r/math/comments/ap25mr/a_monad_is_a_monoid_in_the_category_of/
+
+We can guess a few things, based on the intuition we gained looking at free monoids.
+First, we might note that `return` and `>>=` correspond to `mempty` and `mappend`, respectively.
+Second, we can expect that for any `Functor` `f :: Type -> Type`, there is a corresponding free `Monad` `Free f :: Type -> Type`, the same as for free monoids.
+
+## Free monads
+
+> Basically, the idea behind the Free Monad is that all of our computations become a value.
+> The idea is that whenever we define our application, it does not really execute itself, but constructs an abstract syntax tree that describes the application that we can run later.
+>
+> -- Anatolii Kmetiuk, Mastering Functional Programming
+
+Let us first try to construct `Free f` by analogy.
+We've seen that lists are free monoids.
+The typical definition of a list type in Haskell looks something like this:
+
+```haskell
+data List a = Nil | Cons a (List a)
+```
+
+Now, `List a` is a type, but `Free f` is a unary type constructor, so we'd need one more type argument.
+Doing that directly yields
+
+```haskell
+data Free f a = Nil a | Cons f (Free f) a
+```
+
+We can't define such a data type, due to `f` not being a plain type and hence can't be an argument of a data constructor, but we're actually very close.
+
+Let us now look at the actual definition of `Free f`:
+
+```haskell
+data Free f a = Pure a | Free (f (Free f a))
+```
+
+As you can see, it is indeed very similar to a list -- a bit more general, since the functor is arbitrary, which potentially makes it a tree, with branches of type `f` and leaves of type `a`.
+Pure values, i.e. leaves, are encoded via the constructor `Pure`, and "actions", i.e. branches -- via the constructor `Free`.
+
+It may be helpful to write out the types of these constructors explicitly:
+
+```haskell
+Pure :: a -> Free f a
+Free :: f (Free f a) -> Free f a
+```
+
+If you squint a bit, you will notice that `Pure` has the same signature as `return` (assuming `Monad (Free f)`).
+If you squint some more, you will notice that the signature of `Free` is very similar to the signature of `join`:
```

Comment:
Roughly speaking, `Cons` concatenates the single element and a list of elements, while `join` combines many elements (list of elements for the simplicity). Comparing `Cons` with `Free` looks quite natural to me, as it encodes action followed by the sequence of actions, so that comparing `Free` later with `join` confuses me. I see, that it's about similar signatures, but still feel a bit strange about it. Consequently, I don't get any insight from this analogue.

> this hopefully provides some insight into why such a structure works as an encoding for a free monad

Perhaps, not mentioning `join` would look better __for me__.  Anyway, this is my personal feeling.

Reply:
The point of this comparison is we can define a `Monad` instance with `fmap`, `return` and `join`. We already have `fmap` by the merit of requiring a `Functor` instance, and `Pure` and `Free` provide the encodings for `return` and `join` respectively. Perhaps I should've gone into a bit more detail here.

Reply:
> Comparing Cons with Free looks quite natural to me, as it encodes action followed by the sequence of actions

Also, about that... I briefly mention that `Free` is a bit more general than `Cons`, because, unlike `Cons`, it contains an arbitrary functor, which potentially makes it into a tree rather than a list.

So this intuition of "a sequence of actions" only holds as long as the underlying functor is non-branching, i.e. in terms of Haskell, each data constructor has at most one argument referencing the functor variable. In the general case, `Free` encodes an arbitrary branching tree of actions, not a sequence (which should be evident in the later examples)

Perhaps I should go into a bit more detail in the text.

Reply:
Okay, never mind, I just saw the analogues where they weren't meant.

I've read the paragraph one more time, it sounds pretty clear, though.